File: osp/www/hit.py


import math

from colour import Color


class Hit:

    def __init__(self, hit):

        """
        Set the raw Elasticsearch hit.

        Args:
            hit (dict)
        """

        self.hit = hit

    def field(self, key, delimiter=', '):

        """
        Get a document field.

        Args:
            key (str)

        Returns: str
        """

        value = self.hit['_source'][key]

        if type(value) == list:
            return delimiter.join(value)

        else:
            return value

    def path(self):

        """
        Get the URL path.

        Returns: str
        """

        return '/text/{0}'.format(self.hit['_id'])

    def sort(self):

        """
        Get the sort value.

        Returns: int
        """

        return int(self.hit['sort'][0])

    def color(self, steps=100):

        """
        Get a green -> red scoring color.

        Args:
            steps (int): The number of gradient steps.

        Returns:
            str: A hex color.
        """

        low  = Color('#000000')
        high = Color('#29b730')

        gradient = list(low.range_to(high, steps))
        idx = round(self.field('score')*(steps-1))

        return gradient[idx].get_hex()

File: osp/www/assets/javascripts/ranks/components/hit.js


import _ from 'lodash';
import chroma from 'chroma-js';

import config from './hit.yml';


export default class {


  /**
   * Set the raw hit.
   *
   * @param {Object} hit
   */
  constructor(hit) {
    this.hit = hit;
  }


  /**
   * Get a document field.
   *
   * @param {String} key
   * @return {String}
   */
  field(key, delimiter=', ') {

    // Try to find a highlight.
    let value = (
      _.get(this.hit, `highlight.${key}`) ||
      _.get(this.hit, `_source.${key}`)
    );

    // Join arrays into a string.
    return _.isArray(value) ? value.join(delimiter) : value;

  }


  /**
   * Build the text profile path.
   *
   * @return {String}
   */
  get path() {

    let c = this.hit._source.corpus;
    let i = this.hit._source.identifier;

    return `text/${c}/${i}`;

  }


  /**
   * Get the citation count.
   *
   * @return {Number}
   */
  get count() {
    return this.hit.sort[0].toLocaleString();
  }


  /**
   * Get the overall rank.
   *
   * @return {Number}
   */
  get rank() {
    return this.field('rank').toLocaleString();
  }


  /**
   * Get the teaching score.
   *
   * @return {Number}
   */
  get score() {
    return (this.field('score')*100).toFixed(1);
  }


  /**
   * Get a green -> red scoring color.
   *
   * @return {String}
   */
  get color() {

    let scale = chroma.scale([
      config.colors.low,
      config.colors.high,
    ]).mode('lab');

    return scale(this.field('score')).hex()

  }


}
